### ISSUE TITLE
gtk+ rendering

### ISSUE THREAD

**Issue by w1hkj** [fltk/fltk#443]
OS: Mint Mate 20.3
Platform: i7 core 4

fltk 1.3.5/6/7/8 test/demo application fails to correctly render both Fl_Check_Button and Fl_Round_Button for scheme "gtk+".  Attached file Fl_Light_Button.cxx fixes the bug, but there is probably a better solution.

[Fl_Light_Button.cxx.txt](https://github.com/fltk/fltk/files/8717145/Fl_Light_Button.cxx.txt)
.

**Comment by w1hkj** [comment]
The check box menu item rendering is also an issue.  No visual difference between checked and unchecked on gtk+.

**Comment by Albrecht-S** [comment]
@w1hkj Thank you for the report. As you noticed the gtk+ scheme is special in that it draws the check mark using the standard FL_SELECTION_COLOR rather than the selection color of the widget. Your patch would remove this special behavior.

I looked into it and I'd like to know which default selection color your system selects. Can you please post what `Fl::get_color(FL_SELECTION_COLOR, red, gree, blue)` returns in (uchar) red, green, and blue on your system? Are you maybe using a "dark theme"?

On my Linux Mint Cinnamon I see in test/unittests (Schemes test) a **red** light button and a pretty dark check mark (very likely the system selection color) in the check button below.

**Comment by Albrecht-S** [comment]
PS: the error is related to building "shared libraries" (Windows DLL's). This is something that's still under development. Can you please try either `configure` w/o `--enable-shared` or `CMake` with `OPTION_BUILD_SHARED_LIBS:BOOL=OFF`? This does obviously not build the shared libs but you can continue.

**Comment by w1hkj** [comment]
Built with ./configure --disable-shared.  Checked /usr/local/lib and 
only .a libs present for fltk...

No change in behavior.  Check box and round button both do not display 
the enabled state for the gtk+ scheme.  OK for the other schemes.  
Tested using test/demo.

David

**Comment by Albrecht-S** [comment]
Oh, sorry, my comment above _'... the error is related to building "shared libraries" ...'_ was intended to be posted to another issue. I can't tell why this happened.

However, I'm still interested in answers to the questions in my [earlier comment](https://github.com/fltk/fltk/issues/443#issuecomment-1132844335) about FL_SELECTION_COLOR on your system.

> Tested using test/demo.

Can you please elaborate? `test/demo` is not much more than a menu system that lets you start all the other tests. Which one(s) did you use?

**Comment by w1hkj** [comment]
Sorry Albrecht,

I used the Widgets/Buttons test


	


The same loss of visibility occurs in all applications built on fltk 
when using the gdb+ scheme.

David

**Comment by Albrecht-S** [comment]
David, I still don't know enough about your environment ("Mint Mate 20.3"). Although I'm also using Linux Mint I don't use Mate and I'd like to know more about your setup. Are you maybe using a "dark theme" or something like that?

If possible, can you post a screenshot of the failing test program you're using?

And can you please build and run the following tiny test program and post the output here?
```c++
// Color (theme) test program
// Build: fltk-config --compile color-test.cxx

#include <FL/Fl.H>

int main(int argc, char **argv) {
  uchar red = 0, green = 0, blue = 0;
  Fl::get_system_colors();
  Fl::get_color(FL_SELECTION_COLOR, red, green, blue);
  printf("FL_SELECTION_COLOR   (%3d) = (%02x, %02x, %02x)\n", FL_SELECTION_COLOR, red, green, blue);
  Fl::get_color(FL_BACKGROUND_COLOR, red, green, blue);
  printf("FL_BACKGROUND_COLOR  (%3d) = (%02x, %02x, %02x)\n", FL_BACKGROUND_COLOR, red, green, blue);
  Fl::get_color(FL_BACKGROUND2_COLOR, red, green, blue);
  printf("FL_BACKGROUND2_COLOR (%3d) = (%02x, %02x, %02x)\n", FL_BACKGROUND2_COLOR, red, green, blue);
}
```
For comparison, output on my system (which works fine with the "gtk+" scheme, BTW):
```
./color-test 
FL_SELECTION_COLOR   ( 15) = (00, 00, 80)
FL_BACKGROUND_COLOR  ( 49) = (c0, c0, c0)
FL_BACKGROUND2_COLOR (  7) = (ff, ff, ff)
```

Note: this is only meant to understand your particular color setup better. I'm aware that a real solution would need to change the button drawing code. TIA.

**Comment by Albrecht-S** [comment]
@w1hkj David, I'm currently working on improvements of our `fl_contrast()` function and for further investigations on **this (your) issue** I need my above questions answered. Can you please build and run the small test program I posted [above](https://github.com/fltk/fltk/issues/443#issuecomment-1135850456) and report your results here? Thanks.

**Comment by w1hkj** [comment]
Fedora 36 64 bit Mate desktop executing on HP Chromebook with modified BIOS

Linux CB 5.18.11-200.fc36.x86_64 #1 SMP PREEMPT_DYNAMIC 
Tue Jul 12 22:52:35 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

fltk version: 1.3.8

FL_SELECTION_COLOR   ( 15) = (66, 99, cc)
FL_BACKGROUND_COLOR  ( 49) = (d4, d4, d4)
FL_BACKGROUND2_COLOR (  7) = (d4, d4, d4)


Mint Mate 20.3 64 bit running on Lenovo Desktop M90P

Linux m90p 5.4.0-91-generic #102-Ubuntu SMP 
Fri Nov 5 16:31:28 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

fltk version: 1.3.8

FL_SELECTION_COLOR   ( 15) = (d6, d6, d6)
FL_BACKGROUND_COLOR  ( 49) = (d6, d6, d6)
FL_BACKGROUND2_COLOR (  7) = (d6, d6, d6)

I can repeat on other platforms, but it certainly looks like the color map is system dependent.

David

**Comment by Albrecht-S** [comment]
@w1hkj Thanks for testing and the output. Although I'm not sure how this all plays together [edit: but please see below!], it certainly looks strange if the selection color and the background colors are all the same. 

> I can repeat on other platforms, but it certainly looks like the color map is system dependent.

Sure, but it's not only system dependent, it depends also on your desktop theme and so on, i.e. on personal preferences. Anyway, no need to test further, thanks - for now. It may be interesting though to see why this happens, but that's something maybe for later.

I'll look further into the issue after the `fl_contrast()` stuff is done. My assumption is that we need to change the drawing and use `fl_contrast()` and I'd like to have this fixed before I investigate further.

For a quick test, can you please:
1. revert your source file(s) to the original contents
2. apply the attached tiny patch [light_button.diff](https://github.com/fltk/fltk/files/9151719/light_button.diff.txt)

so we can see if this fixes the issue.

From what I know now (thanks for the test data) it should fix **your** issue but it's maybe _not a complete solution_.

Looking forward to your new test results. TIA.

**Comment by w1hkj** [comment]
Confirm that the patch corrects the visibility of both the check box mark and the round button contrast.

**Comment by Albrecht-S** [comment]
Thanks for testing and for your confirmation.

I'm leaving this issue open until it's finally fixed in the FLTK repo. Please be aware that this will very likely only be fixed in 1.4.0 (no further release of 1.3.x is planned).

Closing STR 3494 now with a link to this issue.

**Comment by Albrecht-S** [comment]
I'm working on a fix, still investigating some special cases.

One question under investigation is how several steps (`Fl::get_system_colors()`, `Fl::scheme("gtk+")`, `Fl_Window::show(argc, argv)` or `Fl_Window::show()` (note: the latter w/o `'argc, argv'`) set the relevant colors. I'm wondering how the OP (@w1hkj) and maybe others (as reported elsewhere, likely in fltk.general) can have system colors where several important FLTK colors are assigned the same value as reported above. Strange...

**_Progress report_**: I have a preliminary fix (similar to the one posted above) but I need to clean up and/or simplify this a bit...

**Comment by Albrecht-S** [comment]
FTR: I updated FLTK 1.3 (branch-1.3) with the minimal patch [light_button.diff](https://github.com/fltk/fltk/files/9151719/light_button.diff.txt) posted earlier. This will be included in release 1.3.9, very likely later this year.

**Comment by Albrecht-S** [comment]
Closing this issue now.

@w1hkj You can still send your confirmation of the changes and/or drop a note if something doesn't work as expected. I'll then reopen the issue.
